lib/plugin-manager.js

```diff
@@ -0,0 +1,303 @@
+/* global Log, Config */
+'use strict';
+
+const EventEmitter = require('events').EventEmitter;
+const async = require('async');
+const StringTemplate = require('./string-template');
+const Metadata = require('./source/metadata');
+const S3 = require('./source/s3');
+
+// const Consul = require('./source/consul');
+// const File = require('./source/file');
+
+const DEFAULT_CALLBACK_DELAY = 2000;
+
+/**
+ * Recursively iterate through an object applying the callback to each k/v pair
+ * @param {Object} obj
+ * @param {Function} callback
+ * @returns {Object}
+ */
+function iter(obj, callback) {
+  const collected = {};
+  let value;
+
+  for (const prop in obj) {
+    if (obj.hasOwnProperty(prop)) {
+      if (typeof obj[prop] === 'object') {
+        value = iter(obj[prop], callback);
+      } else {
+        value = callback(prop, obj[prop]);
+      }
+      collected[prop] = value;
+    }
+  }
+  return collected;
+}
+
+class PluginManager extends EventEmitter {
+  constructor(storage) {
+    super();
+    this.storage = storage;
+    this.index = new S3({
+      interval: Config.get('index:interval'),
+      bucket: Config.get('index:bucket'),
+      path: Config.get('index:path')
+    });
+    this.metadata = new Metadata();
+    this._running = false;
+    this._ok = true;
+    this.updateDelay = DEFAULT_CALLBACK_DELAY;
+    this._boundInit = this._init.bind(this);
+  }
+
+  /**
+   * Initialize the PluginManager's index and metadata source
+   */
+  init() {
+    Log.info('Initializing index and metadata');
+    this._running = true;
+    this._init();
+  }
+
+  /**
+   * Shutdown the plugin manager's sources and then the plugin manager
+   * @param {Boolean} del
```

Comment:
The `del` parameter clears the `Storage` layer's sources array. It was mainly a proposal until we figured out a) how the `PluginManager` shutting down affects its `Storage` object and b) how the `Storage` layer handles clearing its sources.